Assertion failure in ContainerNode::removeNodeWithScriptAssertion via ~PDFPluginTextAnnotation

rniwa opened the pull request:
#### 70f33f17ad48a4dd5defd7a6f7dc72b6764bf8bb
<pre>
Assertion failure in ContainerNode::removeNodeWithScriptAssertion via ~PDFPluginTextAnnotation
<a href="https://bugs.webkit.org/show_bug.cgi?id=250726">https://bugs.webkit.org/show_bug.cgi?id=250726</a>

Reviewed by Chris Dumez.

The debug assertion failure was caused by PDFPluginAnnotation trying to synchronously remove
the annotations. While this is safe since it happens inside a PDF document, delay the work using
the event loop anyway to avoid hitting this assertion.

* LayoutTests/fast/replaced/pdf-iframe-load-crash-expected.txt: Added.
* LayoutTests/fast/replaced/pdf-iframe-load-crash.html: Added.
* LayoutTests/fast/replaced/resources/pdf-iframe-load-crash.pdf: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/dom/Document.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.mm:
(WebKit::PDFPluginAnnotation::~PDFPluginAnnotation):

Canonical link: <a href="https://commits.webkit.org/259033@main">https://commits.webkit.org/259033@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fae0a26cf9bf3a6b122ef65d6f4f979a1d40e400

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103699 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12814 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36644 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112921 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173244 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107649 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13835 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3709 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95941 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112069 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109470 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93729 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38390 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92489 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25335 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80040 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6176 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26727 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6351 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3253 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12332 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46250 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8110 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3298 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.mm:
```diff
@@ -93,7 +94,11 @@
 
-    m_parent->removeChild(*element());
+    m_element->document().eventLoop().queueTask(TaskSource::InternalAsyncTask, [ element = WeakPtr<Node, WeakPtrImplWithEventTargetData> { this->element() } ]() {
+        if (!element)
```
nit: I think this would look a bit nicer:
```
if (RefPtr element = weakElement.get())
    element->remove();
```

rniwa replied:
Ok, sure we can do that.

cdumez reviewed (approved):
Seems OK.

rniwa committed:
Assertion failure in ContainerNode::removeNodeWithScriptAssertion via ~PDFPluginTextAnnotation
https://bugs.webkit.org/show_bug.cgi?id=250726

Reviewed by Chris Dumez.

The debug assertion failure was caused by PDFPluginAnnotation trying to synchronously remove
the annotations. While this is safe since it happens inside a PDF document, delay the work using
the event loop anyway to avoid hitting this assertion.

* LayoutTests/fast/replaced/pdf-iframe-load-crash-expected.txt: Added.
* LayoutTests/fast/replaced/pdf-iframe-load-crash.html: Added.
* LayoutTests/fast/replaced/resources/pdf-iframe-load-crash.pdf: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/dom/Document.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.mm:
(WebKit::PDFPluginAnnotation::~PDFPluginAnnotation):

Canonical link: https://commits.webkit.org/259033@main
